File: reference-implementation/html.js
import { parse } from 'parse5';

/**
 * HTML reading, built on parse5 — the same spec-compliant parser jsdom uses.
 *
 * An earlier version did this with regular expressions. That was wrong for a
 * verification tool: mis-parsed markup (an attribute containing `>`, unclosed
 * tags, CDATA) can mangle a paragraph, which makes a quote that IS on the page
 * read as absent and causes `check` to report drift that does not exist. False
 * alarms are worse than missed evidence, so correctness wins over having no
 * dependencies.
 *
 * Nothing here executes anything. Script, style, noscript, template and svg
 * subtrees are skipped entirely before any text is returned.
 */

const SKIP = new Set(['script', 'style', 'noscript', 'template', 'svg', 'head', 'iframe', 'object', 'canvas']);
const BLOCK = new Set(['h1', 'h2', 'h3', 'h4', 'p', 'li', 'blockquote', 'dd', 'figcaption', 'td', 'th']);
const SECTIONING = new Set(['section', 'article', 'main', 'div', 'header', 'aside', 'nav', 'footer']);

const isElement = n => typeof n.tagName === 'string';
const attr = (node, name) => node.attrs?.find(a => a.name === name)?.value ?? null;
const collapse = s => s.replace(/\s+/gu, ' ').trim();

function* walk(node, ancestors = []) {
  for (const child of node.childNodes ?? []) {
    if (isElement(child)) {
      if (SKIP.has(child.tagName)) continue;
      yield { node: child, ancestors };
      yield* walk(child, [...ancestors, child]);
    } else {
      yield { node: child, ancestors };
    }
  }
}

// Elements that imply a break in the text. Without these, adjacent blocks run
// together ("sign.unclosed") and a quote can appear to span a boundary it never
// crossed.
const SEPARATES = new Set([
  ...BLOCK, ...SECTIONING, 'br', 'hr', 'tr', 'ul', 'ol', 'dl', 'dt', 'table',
  'figure', 'address', 'pre', 'form', 'label', 'option', 'h5', 'h6'
]);

/** Concatenated text of a subtree, skipping non-content elements. */
function textOf(node) {
  if (node.nodeName === '#text') return node.value ?? '';
  if (isElement(node) && SKIP.has(node.tagName)) return '';
  let out = '';
  for (const child of node.childNodes ?? []) {
    if (isElement(child) && SEPARATES.has(child.tagName)) out += ' ';
    out += textOf(child);
    if (isElement(child) && SEPARATES.has(child.tagName)) out += ' ';
  }
  return out;
}

function documentOf(html) {
  return typeof html === 'string' ? parse(html) : html;
}

function findAll(doc, predicate) {
  const out = [];
  for (const { node, ancestors } of walk(doc)) {
    if (isElement(node) && predicate(node)) out.push({ node, ancestors });
  }
  return out;
}

/** All visible text on the page, whitespace-collapsed. */
export function extractText(html) {
  const doc = documentOf(html);
  const body = findAll(doc, n => n.tagName === 'body')[0]?.node;
  return collapse(textOf(body ?? doc));
}

export function extractTitle(html) {
  const t = findAll(documentOf(html), n => n.tagName === 'title')[0]?.node;
  // <title> sits inside <head>, which walk() skips, so read it from the raw tree
  if (t) return collapse(textOf(t)) || null;
  const doc = documentOf(html);
  const found = deepFind(doc, n => isElement(n) && n.tagName === 'title');
  return found ? collapse(textOf(found)) || null : null;
}

function deepFind(node, predicate) {
  for (const child of node.childNodes ?? []) {
    if (predicate(child)) return child;
    const nested = deepFind(child, predicate);
    if (nested) return nested;
  }
  return null;
}

function deepFindAll(node, predicate, out = []) {
  for (const child of node.childNodes ?? []) {
    if (predicate(child)) out.push(child);
    deepFindAll(child, predicate, out);
  }
  return out;
}

/** name= and property= meta tags, keyed by whichever attribute was present. */
export function extractMeta(html) {
  const out = {};
  for (const m of deepFindAll(documentOf(html), n => isElement(n) && n.tagName === 'meta')) {
    const key = attr(m, 'name') ?? attr(m, 'property');
    const content = attr(m, 'content');
    if (key && content != null) out[key.toLowerCase()] = collapse(content);
  }
  return out;
}

/** Parsed application/ld+json blocks. Unparseable blocks are skipped, never guessed at. */
export function extractJsonLd(html) {
  const out = [];
  const scripts = deepFindAll(documentOf(html), n =>
    isElement(n) && n.tagName === 'script' && /application\/ld\+json/i.test(attr(n, 'type') ?? ''));
  for (const s of scripts) {
    const raw = s.childNodes?.map(c => c.value ?? '').join('').trim();
    if (!raw) continue;
    try {
      for (const node of flattenGraph(JSON.parse(raw))) out.push(node);
    } catch { /* malformed: ignored rather than inferred */ }
  }
  return out;
}

function flattenGraph(node) {
  if (Array.isArray(node)) return node.flatMap(flattenGraph);
  if (node && typeof node === 'object') {
    return node['@graph'] ? [node, ...flattenGraph(node['@graph'])] : [node];
  }
  return [];
}

/** Visible text blocks with their own id and nearest sectioning ancestor id. */
export function extractBlocks(html) {
  const doc = documentOf(html);
  const blocks = [];
  for (const { node, ancestors } of walk(doc)) {
    if (!isElement(node) || !BLOCK.has(node.tagName)) continue;
    const text = collapse(textOf(node));
    if (!text) continue;
    let section = null;
    for (let i = ancestors.length - 1; i >= 0; i--) {
      const a = ancestors[i];
      if (SECTIONING.has(a.tagName) && attr(a, 'id')) { section = attr(a, 'id'); break; }
    }
    blocks.push({ tag: node.tagName, id: attr(node, 'id'), section, text });
  }
  return blocks;
}

/**
 * Does this look like a page whose content is assembled in the browser?
 *
 * Fetching alone cannot see client-rendered content. Reporting "no claims
 * found" without saying why leads a publisher to conclude their site has
 * nothing worth publishing, so this is surfaced explicitly.
 */
export function looksClientRendered(html) {
  const doc = documentOf(html);
  const text = extractText(doc);
  const scripts = deepFindAll(doc, n => isElement(n) && n.tagName === 'script');
  const scriptBytes = scripts.reduce((n, s) => n + (attr(s, 'src') ? 0 : (s.childNodes?.[0]?.value?.length ?? 0)), 0);
  const roots = deepFindAll(doc, n => isElement(n) && ['root', 'app', '__next', '__nuxt'].includes(attr(n, 'id') ?? ''));

  const reasons = [];
  if (text.length < 400) reasons.push(`only ${text.length} characters of visible text`);
  if (roots.length && text.length < 1500) reasons.push(`an empty-looking app root (#${attr(roots[0], 'id')})`);
  if (scripts.length > 8 && text.length < 1500) reasons.push(`${scripts.length} script tags but little text`);
  if (scriptBytes > text.length * 4 && text.length < 2000) reasons.push('far more inline script than text');

  return { likely: reasons.length > 0, reasons, textLength: text.length, scriptCount: scripts.length };
}

/** Kept for callers that pass raw strings around; parse5 handles entities itself. */
export function decodeEntities(s) {
  const doc = parse(`<body>${s}</body>`);
  return textOf(deepFind(doc, n => isElement(n) && n.tagName === 'body') ?? doc);
}

File: python/ai_evidence/html.py
"""HTML reading, built on html5lib — a spec-compliant parser.

This mirrors the JavaScript implementation, which uses parse5. Both follow the
HTML5 parsing algorithm, so an attribute containing ``>``, unclosed tags and
implied elements are read the same way in either language.

Regular expressions were tried first and were wrong for a verification tool:
mis-parsed markup makes a quote that IS on the page read as absent, and the
check then reports drift that does not exist.

Nothing here executes anything. Script, style, noscript, template and svg
subtrees are discarded before any text is returned.
"""
from __future__ import annotations

import json
import re
from typing import Any, Iterator

import html5lib

from .patterns import HTML_SETS

SKIP = HTML_SETS["skipElements"]
BLOCK = HTML_SETS["blockElements"]
SECTIONING = HTML_SETS["sectioningElements"]
# Elements that imply a break in the text. Without these, adjacent blocks run
# together and a quote can appear to span a boundary it never crossed.
SEPARATES = BLOCK | SECTIONING | HTML_SETS["separatingElements"]

_WS = re.compile(r"\s+", re.UNICODE)


def _collapse(s: str) -> str:
    return _WS.sub(" ", s).strip()


def _tag(el: Any) -> str:
    t = el.tag
    if not isinstance(t, str):
        return ""
    return t.rsplit("}", 1)[-1].lower()


def _parse(html: str | Any):
    if isinstance(html, str):
        return html5lib.parse(html, namespaceHTMLElements=False)
    return html


def _text_of(el: Any) -> str:
    """Concatenated text of a subtree, skipping non-content elements."""
    name = _tag(el)
    if name in SKIP:
        return ""
    out = el.text or ""
    for child in el:
        if _tag(child) in SKIP:
            out += child.tail or ""
            continue
        sep = " " if _tag(child) in SEPARATES else ""
        out += sep + _text_of(child) + sep + (child.tail or "")
    return out


def _walk(el: Any, ancestors: tuple = ()) -> Iterator[tuple]:
    for child in el:
        if _tag(child) in SKIP:
            continue
        yield child, ancestors
        yield from _walk(child, ancestors + (child,))


def _find(doc: Any, name: str) -> Any:
    for el in doc.iter():
        if _tag(el) == name:
            return el
    return None


def extract_text(html: str | Any) -> str:
    doc = _parse(html)
    body = _find(doc, "body")
    return _collapse(_text_of(body if body is not None else doc))


def extract_title(html: str | Any) -> str | None:
    el = _find(_parse(html), "title")
    if el is None:
        return None
    return _collapse(_text_of(el)) or None


def extract_meta(html: str | Any) -> dict[str, str]:
    """name= and property= meta tags, keyed by whichever attribute was present."""
    out: dict[str, str] = {}
    for el in _parse(html).iter():
        if _tag(el) != "meta":
            continue
        key = el.get("name") or el.get("property")
        content = el.get("content")
        if key and content is not None:
            out[key.lower()] = _collapse(content)
    return out


def _flatten_graph(node: Any) -> list:
    if isinstance(node, list):
        return [x for n in node for x in _flatten_graph(n)]
    if isinstance(node, dict):
        if "@graph" in node:
            return [node] + _flatten_graph(node["@graph"])
        return [node]
    return []


def extract_json_ld(html: str | Any) -> list[dict]:
    """Parsed application/ld+json blocks. Malformed blocks are skipped, never guessed at."""
    out: list[dict] = []
    for el in _parse(html).iter():
        if _tag(el) != "script":
            continue
        if "application/ld+json" not in (el.get("type") or "").lower():
            continue
        raw = (el.text or "").strip()
        if not raw:
            continue
        try:
            out.extend(_flatten_graph(json.loads(raw)))
        except Exception:
            continue  # ignored rather than inferred
    return out


def extract_blocks(html: str | Any) -> list[dict]:
    """Visible text blocks with their own id and nearest sectioning ancestor id."""
    doc = _parse(html)
    blocks = []
    for el, ancestors in _walk(doc):
        name = _tag(el)
        if name not in BLOCK:
            continue
        text = _collapse(_text_of(el))
        if not text:
            continue
        section = None
        for a in reversed(ancestors):
            if _tag(a) in SECTIONING and a.get("id"):
                section = a.get("id")
                break
        blocks.append({"tag": name, "id": el.get("id"), "section": section, "text": text})
    return blocks


def looks_client_rendered(html: str | Any) -> dict:
    """Does this look like a page whose content is assembled in the browser?

    Fetching cannot see client-rendered content. Reporting "no claims found"
    without saying why leads a publisher to conclude their site has nothing
    worth publishing, so the condition is surfaced explicitly.
    """
    doc = _parse(html)
    text = extract_text(doc)
    scripts = [el for el in doc.iter() if _tag(el) == "script"]
    script_bytes = sum(len(el.text or "") for el in scripts if not el.get("src"))
    roots = [el for el in doc.iter() if (el.get("id") or "") in ("root", "app", "__next", "__nuxt")]

    reasons = []
    if len(text) < 400:
        reasons.append(f"only {len(text)} characters of visible text")
    if roots and len(text) < 1500:
        reasons.append(f"an empty-looking app root (#{roots[0].get('id')})")
    if len(scripts) > 8 and len(text) < 1500:
        reasons.append(f"{len(scripts)} script tags but little text")
    if script_bytes > len(text) * 4 and len(text) < 2000:
        reasons.append("far more inline script than text")

    return {"likely": bool(reasons), "reasons": reasons,
            "textLength": len(text), "scriptCount": len(scripts)}
